File: python/variance/ok.py
#!/usr/bin/env python3

# src: chatgpt generated

class WrappedString:
    def __init__(self, value: str):
        self.value = value

    def __str__(self) -> str:
        return self.value

    def __repr__(self) -> str:
        return f'WrappedString({self.value})'

    def upper(self) -> 'WrappedString':
        return WrappedString(self.value.upper())

    def lower(self) -> 'WrappedString':
        return WrappedString(self.value.lower())

    def capitalize(self) -> 'WrappedString':
        return WrappedString(self.value.capitalize())

    # Define other methods as needed
class ReversedString(WrappedString):
    def __str__(self) -> str:
        return self.value[::-1]

class TitleString(WrappedString):
    def __str__(self) -> str:
        return self.value.title()

class Animal:
    def make_sound(self) -> WrappedString:
        return WrappedString("Generic animal sound")

class Dog(Animal):
    def make_sound(self) -> ReversedString:
        return ReversedString("Woof!")

class Cat(Animal):
    def make_sound(self) -> TitleString:
        return TitleString("Meow!")

class Bird(Animal):
    def make_sound(self) -> WrappedString:
        return WrappedString("Tweet!")

def print_sound(animal: Animal) -> None:
    print(animal.make_sound())

dog = Dog()
cat = Cat()
bird = Bird()

#so by being able to pass Dog/Cat/Bird to print_sound which expects an Animal(aka base class type) we're doing contravariance
#and covariance is that make_sound() is expected to return a WrappedString (from Animal) type, but we can have it return other subtypes of WrappedString.
print_sound(dog)   # Covariance: Dog's make_sound returns ReversedString
print_sound(cat)   # Covariance: Cat's make_sound returns TitleString
print_sound(bird)  # Covariance: Bird's make_sound returns WrappedString

File: python/variance/ok.rs
// Define a wrapper trait for WrappedString that erases the size of Self
trait WrappedStringDyn {
    fn to_string_dyn(&self) -> String;
}

// Implement WrappedStringDyn for any type that implements WrappedString
impl<T: WrappedString> WrappedStringDyn for T {
    fn to_string_dyn(&self) -> String {
        self.to_string()
    }
}

// Define the original trait for wrapped strings
trait WrappedString {
    fn new(value: &str) -> Self;
    fn to_string(&self) -> String;
}

// Implement the trait for a simple string wrapper
struct SimpleString(String);

impl WrappedString for SimpleString {
    fn new(value: &str) -> Self {
        SimpleString(value.to_string())
    }

    fn to_string(&self) -> String {
        self.0.clone()
    }
}

// Implement the trait for a reversed string wrapper
struct ReversedString(String);

impl WrappedString for ReversedString {
    fn new(value: &str) -> Self {
        ReversedString(value.chars().rev().collect())
    }

    fn to_string(&self) -> String {
        self.0.clone()
    }
}

// Implement the trait for a title-cased string wrapper
struct TitleString(String);

impl WrappedString for TitleString {
    fn new(value: &str) -> Self {
        TitleString(value.to_string())
    }

    fn to_string(&self) -> String {
        self.0.clone()
    }
}

// Define an enum to represent different types of animals
enum AnimalType {
    Dog,
    Cat,
    Bird,
}

// Define a trait for animals
trait Animal {
    fn make_sound(&self) -> Box<dyn WrappedStringDyn>;
}

// Implement the trait for each type of animal
struct Dog;
impl Animal for Dog {
    fn make_sound(&self) -> Box<dyn WrappedStringDyn> {
        Box::new(ReversedString::new("Woof!"))
    }
}

struct Cat;
impl Animal for Cat {
    fn make_sound(&self) -> Box<dyn WrappedStringDyn> {
        Box::new(TitleString::new("Meow!"))
    }
}

struct Bird;
impl Animal for Bird {
    fn make_sound(&self) -> Box<dyn WrappedStringDyn> {
        Box::new(SimpleString::new("Tweet!"))
    }
}

// Function to print the sound of an animal
fn print_sound(animal: &dyn Animal) {
    let sound = animal.make_sound();
    println!("{}", sound.to_string_dyn());
}

fn main() {
    let dog = Dog;
    let cat = Cat;
    let bird = Bird;

    print_sound(&dog);   // Output: foow
    print_sound(&cat);   // Output: Meow!
    print_sound(&bird);  // Output: Tweet!
}

